ATM90E32 Semi-automatic calibration & Status fields

CircuitSetup opened the pull request:
# What does this implement/fix?

This PR adds support for semi-automatic gain and offset calibration, and status messages per phase to the ATM90E32 power metering component.

Features added:
- Gain calibration and offset calibration routines
- Reference voltage and current input via `number:` fields
- Clear buttons for all calibration types
- New text sensors for reporting phase and frequency status - over voltage, voltage sag, over current, phase loss, and frequency low/high
- Flash-based storage for gain, offset, and power offset values as well as new fields to store them in the config file
- Finished adding apparent power sensor

Fixes:
- get_phase_reverse_active_energy_ only getting values for Phase A
- "SPI power factor read error." appearing when power factor was a negative number
- Typo with LOG_SENSOR returning Harmonic Power, Phase Angle, and Peak Current for A, when it should've been B, and C

Improvements:
- Moved get sensor data in main loop into 1 for loop instead of 19
- In read16_, reduced delay_microseconds_safe from 10 to 1 since the min time between CS low and SCK 200ns, and 1ms is fine and probably not even needed

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- esphome/issues/issues/3741

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#4810

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
spi:
    clk_pin: 18
    miso_pin: 19
    mosi_pin: 23

sensor:
  - platform: atm90e32
    cs_pin: 5
    id: atm90e32_chip1
    phase_a:
      voltage:
        name: EMON Line Voltage A
      current:
        name: EMON CT1 Current
      power:
        name: EMON Active Power CT1
      reactive_power:
        name: EMON Reactive Power CT1
      apparent_power:
        name: EMON Apparent Power CT1
      harmonic_power:
        name: EMON Harmonic Power CT1
      power_factor:
        name: EMON Power Factor CT1
      phase_angle:
        name: EMON Phase Angle CT1
      peak_current:
        name: EMON Peak Current CT1
      gain_voltage: 7305
      gain_ct: 27961
      offset_voltage: 0
      offset_current: 0
      offset_active_power: 0
      offset_reactive_power: 0
    phase_b:
      current:
        name: EMON CT2 Current
      power:
        name: EMON Active Power CT2
      reactive_power:
        name: EMON Reactive Power CT2
      apparent_power:
        name: EMON Apparent Power CT2
      harmonic_power:
        name: EMON Harmonic Power CT2
      power_factor:
        name: EMON Power Factor CT2
      phase_angle:
        name: EMON Phase Angle CT2
      peak_current:
        name: EMON Peak Current CT2
      gain_voltage: 7305
      gain_ct: 27961
      offset_voltage: 0
      offset_current: 0
      offset_active_power: 0
      offset_reactive_power: 0
    phase_c:
      current:
        name: EMON CT3 Current
      power:
        name: EMON Active Power CT3
      reactive_power:
        name: EMON Reactive Power CT3
      apparent_power:
        name: EMON Apparent Power CT3
      harmonic_power:
        name: EMON Harmonic Power CT3
      power_factor:
        name: EMON Power Factor CT3
      phase_angle:
        name: EMON Phase Angle CT3
      peak_current:
        name: EMON Peak Current CT3
      gain_voltage: 7305
      gain_ct: 27961
      offset_voltage: 0
      offset_current: 0
      offset_active_power: 0
      offset_reactive_power: 0
    frequency:
      name: EMON Line Frequency
    chip_temperature:
      name: EMON Chip Temp
    line_frequency: 60Hz
    current_phases: 3
    gain_pga: 1X
    enable_offset_calibration: True
    enable_gain_calibration: True

text_sensor:
  - platform: atm90e32
    id: atm90e32_chip1
    phase_status:
      phase_a:
        name: "Phase A Status"
      phase_b:
        name: "Phase B Status"
      phase_c:
        name: "Phase C Status"
    freq_status:
      name: "Frequency Status"

button:
  - platform: atm90e32
    id: atm90e32_chip1
    run_gain_calibration:
      name: "Run Gain Calibration"
    clear_gain_calibration:
      name: "Clear Gain Calibration"
    run_offset_calibration:
      name: "Run Offset Calibration"
    clear_offset_calibration:
      name: "Clear Offset Calibration"
    run_power_offset_calibration:
      name: "Run Power Offset Calibration"
    clear_power_offset_calibration:
      name: "Clear Power Offset Calibration"

number:
  - platform: atm90e32
    id: atm90e32_chip1
    reference_voltage:
      phase_a:
        name: "Phase A Ref Voltage"
      phase_b:
        name: "Phase B Ref Voltage"
      phase_c:
        name: "Phase C Ref Voltage"
    reference_current:
      phase_a:
        name: "Phase A Ref Current"
      phase_b:
        name: "Phase B Ref Current"
      phase_c:
        name: "Phase C Ref Current"
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

CircuitSetup committed:
Calibrations & status functions, bugfixes, and improvements

- added function run_gain_calibration - calculates gain_voltage and gain_ct based on a reference value number field
- added save_gain_calibration_to_memory - saves gain calibrations to flash memory
- added run_power_offset_calibrations, calibrate_power_offset_ - calculates active and reactive power offsets
- modified calibrate_voltage_offset_phase & calibrate_current_offset_phase, and combined them into calibrate_offset_ since the calculations are the same
- added write_gains_to_registers, write_offsets_to_registers, write_power_offsets_to_register - takes calibration calculation values and writes them to the corresponding registers
- added restore_gain_calibrations, restore_offset_calibrations, restore_power_offset_calibrations - restores calibration calculations after a reboot, and writes them to the proper registers, if they exist
- added clear_gain_calibrations, clear_offset_calibrations, clear_power_offset_calibrations - triggered by button input, clears calibration calculations from registers, and either writes the default values to the proper registers, or uses values from the config file if they exist
- added check_phase_status - returns if a phase has a flag set for "Over Voltage", "Voltage Sag", "Phase Loss", otherwise "Okay"
- added check_over_current - checks if a phase is over the register threshold of 65.53A
- added calculate_voltage_threshold - calculates the values that are written to the SAGTH (voltage sag) and OVTH (over voltage) registers based on line_freq (60Hz = 120V, 50Hz = 220V)
- added check_freq_status - returns if an atm90e32 has a flag set for "HIGH" or "LOW" frequency, otherwise "Normal"
- added get_phase_apparent_power - gets data from the apparent power registers
- fixed bug with get_phase_reverse_active_energy_ only getting values for Phase A
- fixed bug with "SPI power factor read error." appearing when power factor was a negative number
- fixed typo with LOG_SENSOR returning Harmonic Power, Phase Angle, and Peak Current for A, when it should've been B, and C
- improvement - put get sensor data in main loop in 1 for loop instead of 19
- improvement - in read16_, reduced delay_microseconds_safe from 10 to 1 since the min time between CS low and SCK 200ns, and 1ms is fine and probably not even needed

CircuitSetup committed:
add power offset, gain calibration, and clear buttons

CircuitSetup committed:
added fields to store offsets, enable statuses in config file

- added voltage_offset, current_offset, active_power_offset, reactive_power_offset
- added enable_gain_calibration, phase_status, freq_status
- changed default gain_pga to 1x, because that's what it should be for most CTs
- added name to atm90e32 sensor
- changed CONF_APPARENT_POWER, unit_of_measurement to UNIT_VOLT_AMPS
- added device_class=DEVICE_CLASS_POWER, to CONF_REACTIVE_POWER

CircuitSetup committed:
add reference voltage/current input number fields

CircuitSetup committed:
add status fields

CircuitSetup committed:
update test

CircuitSetup committed:
Update ci.yml

CircuitSetup committed:
clangd & ruff formatting

CircuitSetup committed:
lint & flake 8 errors

CircuitSetup committed:
clang-tidy

CircuitSetup committed:
clang-tidy

CircuitSetup committed:
Update atm90e32_button.h

CircuitSetup committed:
clang-tidy again

CircuitSetup committed:
Update atm90e32_button.h

CircuitSetup committed:
clang-tidy does not like this

CircuitSetup committed:
Update atm90e32_button.h

CircuitSetup committed:
Update ci.yml

CircuitSetup committed:
Update ci.yml

descipher commented:
Awesome John, I will do some review and testing asap. Probably next week or so.

CircuitSetup committed:
Merge branch 'dev' into atm90e32-calibration-status

CircuitSetup committed:
Fixed checks for run offset calibrations

also removed redundant get_enable_gain_calibration, and check in the button function

CircuitSetup committed:
clangd formatting

CircuitSetup committed:
clangd

jesserockz commented on esphome/components/atm90e32/text_sensor/__init__.py:
```diff
@@ -0,0 +1,64 @@
+    parent = await cg.get_variable(config[CONF_ID])
+
+    if CONF_PHASE_STATUS in config:
+        phase_cfg = config[CONF_PHASE_STATUS]
```
```suggestion
    if phase_cfg := config.get(CONF_PHASE_STATUS):
```

jesserockz commented on esphome/components/atm90e32/text_sensor/__init__.py:
```diff
@@ -0,0 +1,64 @@
+        phase_cfg = config[CONF_PHASE_STATUS]
+        for i, key in enumerate(PHASE_KEYS):
+            if key in phase_cfg:
+                sens = await text_sensor.new_text_sensor(phase_cfg[key])
```
```suggestion
            if sub_phase_cfg := phase_cfg.get(key):
                sens = await text_sensor.new_text_sensor(sub_phase_cfg)
```

jesserockz commented on esphome/components/atm90e32/text_sensor/__init__.py:
```diff
@@ -0,0 +1,64 @@
+                cg.add(parent.set_phase_status_text_sensor(i, sens))
+
+    if CONF_FREQ_STATUS in config:
+        sens = await text_sensor.new_text_sensor(config[CONF_FREQ_STATUS])
```
```suggestion
    if freq_status_config := config.get(CONF_FREQ_STATUS):
        sens = await text_sensor.new_text_sensor(freq_status_config)
```

jesserockz commented on esphome/components/atm90e32/text_sensor/__init__.py:
```diff
@@ -0,0 +1,64 @@
+)
+
+CONF_PHASE_STATUS = "phase_status"
+CONF_FREQ_STATUS = "freq_status"
```
Longer, proper names are preferred
```suggestion
CONF_FREQUENCY_STATUS = "frequency_status"
```

jesserockz commented on esphome/components/atm90e32/sensor.py:
```diff
@@ -137,13 +149,18 @@
     cv.Schema(
         {
             cv.GenerateID(): cv.declare_id(ATM90E32Component),
+            cv.Optional(CONF_NAME): cv.string,
```
Not used?
```suggestion
```

jesserockz commented on esphome/components/atm90e32/number/__init__.py:
```diff
@@ -0,0 +1,137 @@
+    return cv.All(
+        cv.Schema(
+            {
+                cv.Required(CONF_NAME): cv.string,
```
Name is already provided by the `number_schema`
```suggestion
```

jesserockz commented on esphome/components/atm90e32/number/__init__.py:
```diff
@@ -0,0 +1,137 @@
+                cv.Optional("optimistic", default=True): cv.boolean,
+                cv.Optional("min_value", default=1.0): cv.float_,
+                cv.Optional("max_value", default=200.0): cv.float_,
+                cv.Optional("step", default=0.1): cv.float_,
```
constants must be used for all config keys

jesserockz commented on esphome/components/atm90e32/number/__init__.py:
```diff
@@ -0,0 +1,137 @@
+
+
+def reference_current_schema():
+    return cv.All(
```
This doesnt need to be a method, it can be static
```suggestion
REFERENCE_CURRENT_SCHEMA = cv.All(
```

jesserockz commented on esphome/components/atm90e32/number/__init__.py:
```diff
@@ -0,0 +1,137 @@
+            {
+                cv.Required(CONF_NAME): cv.string,
+                cv.Optional("mode", default="box"): cv.string,
+                cv.Optional("optimistic", default=True): cv.boolean,
```
Not used?
```suggestion
```

CircuitSetup replied:
the reference value isn't published to a hardware register, so no confirmation is wanted in this case

jesserockz commented on esphome/components/atm90e32/number/__init__.py:
```diff
@@ -0,0 +1,137 @@
+async def to_code(config):
+    parent = await cg.get_variable(config[CONF_ID])
+
+    if CONF_REFERENCE_VOLTAGE in config:
+        voltage_cfg = config[CONF_REFERENCE_VOLTAGE]
```
```suggestion
    if voltage_cfg := config.get(CONF_REFERENCE_VOLTAGE):
```

jesserockz commented on esphome/components/atm90e32/number/__init__.py:
```diff
@@ -0,0 +1,137 @@
+        voltage_objs = [None, None, None]
+
+        for i, key in enumerate(PHASE_KEYS):
+            if key in voltage_cfg:
+                validated = reference_voltage_schema()(voltage_cfg[key])
```
Config is already validated before to_code is called
```suggestion
            if validated := voltage_cfg.get(key):
```

jesserockz commented on esphome/components/atm90e32/number/__init__.py:
```diff
@@ -0,0 +1,137 @@
+                if voltage_objs[i] is None:
+                    voltage_objs[i] = voltage_objs[0]
+
+        for i in range(3):
+            if voltage_objs[i] is not None:
+                cg.add(parent.set_reference_voltage(i, voltage_objs[i]))
```
```suggestion
        for i, obj in enumerate(voltage_objs):
            if obj is not None:
                cg.add(parent.set_reference_voltage(i, obj))
```

jesserockz commented on esphome/components/atm90e32/number/__init__.py:
```diff
@@ -0,0 +1,137 @@
+            if voltage_objs[i] is not None:
+                cg.add(parent.set_reference_voltage(i, voltage_objs[i]))
+
+    if CONF_REFERENCE_CURRENT in config:
+        current_cfg = config[CONF_REFERENCE_CURRENT]
```
```suggestion
    if current_cfg := config.get(CONF_REFERENCE_CURRENT):
```

jesserockz commented on esphome/components/atm90e32/number/__init__.py:
```diff
@@ -0,0 +1,137 @@
+    if CONF_REFERENCE_CURRENT in config:
+        current_cfg = config[CONF_REFERENCE_CURRENT]
+        for i, key in enumerate(PHASE_KEYS):
+            if key in current_cfg:
+                validated = reference_current_schema()(current_cfg[key])
```
As above
```suggestion
            if validate := current_cfg.get(key):
```

jesserockz commented on esphome/components/atm90e32/number/__init__.py:
```diff
@@ -0,0 +1,137 @@
+PHASE_KEYS = [CONF_PHASE_A, CONF_PHASE_B, CONF_PHASE_C]
+
+
+def reference_voltage_schema():
```
Please apply the same changes as current below

jesserockz commented on esphome/components/atm90e32/sensor.py:
```diff
@@ -33,22 +34,32 @@
 
 from . import atm90e32_ns
 
+ATM90E32Number = atm90e32_ns.class_("ATM90E32Number", number.Number)
+
```
Not used
```suggestion
```

jesserockz commented on esphome/components/atm90e32/number/__init__.py:
```diff
@@ -0,0 +1,137 @@
+
+CONFIG_SCHEMA = cv.Schema(
+    {
+        cv.Required(CONF_ID): cv.use_id(ATM90E32Component),
```
This can automatically attach if there is only one parent in config
```suggestion
        cv.GenerateID(CONF_ID): cv.use_id(ATM90E32Component),
```

jesserockz commented on esphome/components/atm90e32/button/__init__.py:
```diff
@@ -1,43 +1,95 @@
 
 CONFIG_SCHEMA = {
-    cv.GenerateID(CONF_ID): cv.use_id(ATM90E32Component),
+    cv.Required(CONF_ID): cv.use_id(ATM90E32Component),
```
```suggestion
    cv.GeneratedID(CONF_ID): cv.use_id(ATM90E32Component),
```

jesserockz commented on esphome/components/atm90e32/atm90e32.h:
```diff
@@ -119,21 +188,50 @@ class ATM90E32Component : public PollingComponent,
+  bool enable_gain_calibration_{false};
+};
+
+class ATM90E32Number : public number::Number, public Parented<ATM90E32Component> {
+ public:
+  void control(float value) override { this->publish_state(value); }
+};
```
This should be in `atm90e32/number/atm90e32_number.h`

jesserockz commented on esphome/components/atm90e32/atm90e32.h:
```diff
@@ -119,21 +188,50 @@ class ATM90E32Component : public PollingComponent,
+ public:
+  void control(float value) override { this->publish_state(value); }
+};
+class ATM90E32PhaseStatusSensor : public text_sensor::TextSensor, public Parented<ATM90E32Component> {
+ public:
+  void set_parent(ATM90E32Component *parent) { this->parent_ = parent; }
+};
+
+class ATM90E32FreqStatusSensor : public text_sensor::TextSensor, public Parented<ATM90E32Component> {
+ public:
+  void set_parent(ATM90E32Component *parent) { this->parent_ = parent; }
```
This should be in `atm90e32/text_sensor/atm90e32_text_sensor.h`

jesserockz commented on esphome/components/atm90e32/atm90e32.h:
```diff
@@ -119,21 +188,50 @@ class ATM90E32Component : public PollingComponent,
+  ESPPreferenceObject gain_calibration_pref_;
 
   sensor::Sensor *freq_sensor_{nullptr};
+  text_sensor::TextSensor *phase_status_text_sensor_[3]{nullptr};
+  text_sensor::TextSensor *freq_status_text_sensor_{nullptr};
```
Any code like this that touches any other platform needs to be wrapped in `ifdefs`
This applies to any includes, references like here and code that touches those references.

```suggestion
#ifdef USE_TEXT_SENSOR
  text_sensor::TextSensor *phase_status_text_sensor_[3]{nullptr};
  text_sensor::TextSensor *freq_status_text_sensor_{nullptr};
#endif
```

jesserockz commented on esphome/components/atm90e32/atm90e32.h:
```diff
@@ -51,53 +82,91 @@ class ATM90E32Component : public PollingComponent,
+  uint16_t calculate_voltage_threshold(int line_freq, uint16_t ugain, float multiplier);
   int32_t last_periodic_millis = millis();
 
+ private:
```
This should be protected, we generally dont use private in ESPHome

jesserockz commented on esphome/components/atm90e32/atm90e32.h:
```diff
@@ -51,53 +82,91 @@ class ATM90E32Component : public PollingComponent,
   int32_t last_periodic_millis = millis();
 
+ private:
+  number::Number *ref_voltages_[3]{nullptr, nullptr, nullptr};
+  number::Number *ref_currents_[3]{nullptr, nullptr, nullptr};
```
Same as the text sensor stuff
```suggestion
#ifdef USE_NUMBER
  number::Number *ref_voltages_[3]{nullptr, nullptr, nullptr};
  number::Number *ref_currents_[3]{nullptr, nullptr, nullptr};
#endif
```

jesserockz commented on esphome/components/atm90e32/atm90e32.cpp:
```diff
@@ -223,27 +138,79 @@ void ATM90E32Component::setup() {
+  if (this->enable_offset_calibration_) {
+    // Initialize flash storage for offset calibrations
+    uint32_t o_hash =
+        fnv1_hash(App.get_friendly_name() + "_offset_calibration_" + std::to_string((intptr_t) this->cs_));
```
Not sure I like using cs here for this..

jesserockz reviewed (changes requested):
I have left a lot of suggestions below and there are probably more but I have run out of time today and decided to just submit this as is for now.

Considering the number of platforms for this component has increased quite a bit, maybe consider implementing the base component with a top level component eg.
```yaml
atm90e32:
  id: some_id
  cs_pin: GPIOXX
  ... # other common config

sensor:
  - platform: atm90e32
    atm90e32_id: some_id
    ...

...
```

It will be a breaking change, but it will help clean up some of the code too and allow moving things into more specific files.

Please ping me on discord in devs if you have questions about this as I get way too many GH notifications to notice.

Jesse

CircuitSetup committed:
add ifdefs, move component declarations to .h files

... and don't use to_string with pointer in hashes, as this could easily change

CircuitSetup committed:
clean up python files

- removed name from atm90e32 sensor
- "freq_status" to "frequency_status"
- don't need cv.Required(CONF_NAME): cv.string, in reference number
- constants used for config keys in number
- used static variables for reference number config schemas
- for number, config is already validated before to_code is called
- for number & button, cv.GenerateID(CONF_ID): cv.use_id(ATM90E32Component), instead of cv.Required

Co-Authored-By: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

CircuitSetup committed:
update frequency_status:

CircuitSetup committed:
lint newline at end

CircuitSetup commented:
> I have left a lot of suggestions below and there are probably more but I have run out of time today and decided to just submit this as is for now.
>
> Considering the number of platforms for this component has increased quite a bit, maybe consider implementing the base component with a top level component eg.

As discussed, I rather leave it as-is for now.

All other changes have been put in

CircuitSetup committed:
Improved logging output, bugs, & cleanup

- run_gain_calibrations now accounts for cases where reference values or measured values are 0 independently of each other and output log messages accordingly
- fixed run_gain_calibrations writing new voltage/current gains twice
- added helper function validate_spi_read_
- fixed get_phase_angle return casting before the ternary
- fixed get_phase_peak_current_ unnecessary (float) cast
- restore_gain_calibrations_ outputs what it's restoring

CircuitSetup committed:
Update atm90e32.cpp

default doesn't go there...

CircuitSetup committed:
Update sensor.py

correct frequency status

jesserockz committed:
Fix some cross platform dependencies

CircuitSetup committed:
Merge branch 'dev' into atm90e32-calibration-status

jesserockz commented on esphome/components/atm90e32/button/atm90e32_button.cpp:
```diff
@@ -1,20 +1,78 @@
 namespace atm90e32 {
 
-static const char *const TAG = "atm90e32.button";
+static const char *const TAG = "CALIBRATION Button";
```
This should not change
```suggestion
static const char *const TAG = "atm90e32.button";
```

jesserockz committed:
Update esphome/components/atm90e32/button/atm90e32_button.cpp

jesserockz commented on esphome/components/atm90e32/number/__init__.py:
```diff
@@ -0,0 +1,133 @@
+    cv.Schema(
+        {
+            cv.Optional(CONF_MODE, default="box"): cv.string,
+            cv.Optional(CONF_OPTIMISTIC, default=True): cv.boolean,
```
Not used
```suggestion
```

jesserockz commented on esphome/components/atm90e32/number/__init__.py:
```diff
@@ -0,0 +1,133 @@
+    cv.Schema(
+        {
+            cv.Optional(CONF_MODE, default="box"): cv.string,
+            cv.Optional(CONF_OPTIMISTIC, default=True): cv.boolean,
```
Not used
```suggestion
```

jesserockz commented on esphome/components/atm90e32/number/__init__.py:
```diff
@@ -0,0 +1,133 @@
+    CONF_MAX_VALUE,
+    CONF_MIN_VALUE,
+    CONF_MODE,
+    CONF_OPTIMISTIC,
```
```suggestion
```

jesserockz commented on esphome/components/atm90e32/text_sensor/__init__.py:
```diff
@@ -0,0 +1,63 @@
+
+CONFIG_SCHEMA = cv.Schema(
+    {
+        cv.Required(CONF_ID): cv.use_id(ATM90E32Component),
```
This way it can be optional if there is only one main sensor entry
```suggestion
        cv.GenerateID(): cv.use_id(ATM90E32Component),
```

jesserockz commented on esphome/components/atm90e32/text_sensor/atm90e32_text_sensor.h:
You actually don't even need these sub classes. sensors and text sensors can be created without a specific class, and is not needed if the class is not doing something specific.

This file can be deleted and changes below applied

jesserockz commented on esphome/components/atm90e32/text_sensor/__init__.py:
```diff
@@ -0,0 +1,63 @@
+    "ATM90E32FreqStatusSensor",
+    text_sensor.TextSensor,
+    cg.Parented.template(ATM90E32Component),
+)
```
```suggestion
```

jesserockz commented on esphome/components/atm90e32/text_sensor/__init__.py:
```diff
@@ -0,0 +1,63 @@
+        ),
+        cv.Optional(CONF_PHASE_C): text_sensor.text_sensor_schema(
+            ATM90E32PhaseStatusSensor, icon="mdi:flash-alert"
+        ),
```
```suggestion
        cv.Optional(CONF_PHASE_A): text_sensor.text_sensor_schema(
            icon="mdi:flash-alert"
        ),
        cv.Optional(CONF_PHASE_B): text_sensor.text_sensor_schema(
            icon="mdi:flash-alert"
        ),
        cv.Optional(CONF_PHASE_C): text_sensor.text_sensor_schema(
            icon="mdi:flash-alert"
        ),
```

jesserockz commented on esphome/components/atm90e32/text_sensor/__init__.py:
```diff
@@ -0,0 +1,63 @@
+        cv.Required(CONF_ID): cv.use_id(ATM90E32Component),
+        cv.Optional(CONF_PHASE_STATUS): PHASE_STATUS_SCHEMA,
+        cv.Optional(CONF_FREQUENCY_STATUS): text_sensor.text_sensor_schema(
+            ATM90E32FreqStatusSensor, icon="mdi:lightbulb-alert"
```
```suggestion
            icon="mdi:lightbulb-alert"
```

jesserockz commented on esphome/components/atm90e32/text_sensor/__init__.py:
```diff
@@ -0,0 +1,63 @@
+        for i, key in enumerate(PHASE_KEYS):
+            if sub_phase_cfg := phase_cfg.get(key):
+                sens = await text_sensor.new_text_sensor(sub_phase_cfg)
+                await cg.register_parented(sens, parent)
```
```suggestion
```

jesserockz commented on esphome/components/atm90e32/text_sensor/__init__.py:
```diff
@@ -0,0 +1,63 @@
+
+    if freq_status_config := config.get(CONF_FREQUENCY_STATUS):
+        sens = await text_sensor.new_text_sensor(freq_status_config)
+        await cg.register_parented(sens, parent)
```
```suggestion
```

jesserockz commented on esphome/components/atm90e32/atm90e32.cpp:
```diff
@@ -467,29 +440,426 @@ float ATM90E32Component::get_chip_temperature_() {
-uint16_t ATM90E32Component::calibrate_voltage_offset_phase(uint8_t phase) {
+void ATM90E32Component::run_gain_calibrations() {
+  if (!this->enable_gain_calibration_) {
+    ESP_LOGW("CALIBRATION", "Gain calibration is disabled! Enable it first with enable_gain_calibration: true");
```
Logs should always use the TAG please. If you want to prefix with something then you just need to add it to the beginning of the message string. 

If you do want a different TAG for the calibration stuff, you can move these functions into their own c++ file with its own TAG eg `atm90e32_calibration.cpp` with `... TAG = "atm90e32.calibration"`

```suggestion
    ESP_LOGW(TAG, "[CALIBRATION] Gain calibration is disabled! Enable it first with enable_gain_calibration: true");
```

jesserockz reviewed (changes requested):
I think this should be the final round of changes. If done by early next week, it will be included in the next release.

CircuitSetup committed:
fix Log TAG, remove text_sensor sub classes...

remove CONF_OPTIMISTIC from number

CircuitSetup committed:
Merge branch 'atm90e32-calibration-status' of https://github.com/CircuitSetup/esphome into atm90e32-calibration-status

CircuitSetup committed:
Update __init__.py

jesserockz reviewed (approved):
Thanks for the quick turnaround.
